ext/datadog_profiling_native_extension/collectors_thread_context.c

```diff
@@ -1551,6 +1613,124 @@ static VALUE _native_sample_skipped_allocation_samples(DDTRACE_UNUSED VALUE self
+// otherwise undisturbed).
+//
+// Specifically, the way this works is:
+// 1. The latest entry in the opentelemetry context storage represents the current span (if any). We take the span id
+//    and trace id from this span.
+// 2. To find the local root span id, we walk the context storage backwards from the current span, and find the earliest
+//    entry in the context storage that has the same trace id as the current span; we use the found span as the local
+//    root span id.
+//    This matches the semantics of how ddtrace tracing creates a TraceOperation and assigns a local root span to it.
+static void otel_without_ddtrace_trace_identifiers_for(
+  struct thread_context_collector_state *state,
+  VALUE thread,
+  struct trace_identifiers *trace_identifiers_result
+) {
+  VALUE context_storage = rb_thread_local_aref(thread, otel_context_storage_id /* __opentelemetry_context_storage__ */);
+
+  // If it exists, context_storage is expected to be an Array[OpenTelemetry::Context]
+  if (context_storage == Qnil || !RB_TYPE_P(context_storage, T_ARRAY)) return;
+
+  VALUE otel_current_span_key = get_otel_current_span_key(state);
+  if (otel_current_span_key == Qfalse) return;
+
+  int active_context_index = RARRAY_LEN(context_storage) - 1;
+  if (active_context_index < 0) return;
+
+  struct otel_span active_span_and_context = otel_span_and_context_from(rb_ary_entry(context_storage, active_context_index), otel_current_span_key);
+  // If it exists, active_span_context is expected to be a OpenTelemetry::Trace::SpanContext (don't confuse it with OpenTelemetry::Context)
+  VALUE active_span_context = active_span_and_context.span_context;
+  if (active_span_context == Qnil) return;
+
+  // Get the span id and trace id from the active span...
+  VALUE active_span_id = rb_ivar_get(active_span_context, at_span_id_id /* @span_id */);
+  VALUE active_span_trace_id = rb_ivar_get(active_span_context, at_trace_id_id /* @trace_id */);
+  if (active_span_id == Qnil || active_span_trace_id == Qnil || !RB_TYPE_P(active_span_id, T_STRING) || !RB_TYPE_P(active_span_trace_id, T_STRING)) return;
+
+  VALUE local_root_span_id = active_span_id;
+  VALUE local_root_span = active_span_and_context.span;
+
+  // Now find the oldest span starting from the active span that still has the same trace id as the active span
+  for (int i = active_context_index - 1; i >= 0; i--) {
+    struct otel_span span_and_context = otel_span_and_context_from(rb_ary_entry(context_storage, i), otel_current_span_key);
+    VALUE span_context = span_and_context.span_context;
+    if (span_context == Qnil) return;
+
+    VALUE span_id = rb_ivar_get(span_context, at_span_id_id /* @span_id */);
+    VALUE span_trace_id = rb_ivar_get(span_context, at_trace_id_id /* @trace_id */);
+    if (span_id == Qnil || span_trace_id == Qnil || !RB_TYPE_P(span_id, T_STRING) || !RB_TYPE_P(span_trace_id, T_STRING)) return;
```

Comment:
Every time you call `otel_span_and_context_from` you end up doing these things (refer to lines 1655-1661 above).

Could we not include all of these into `otel_span_and_context_from`? It seems like it'd improve readability and maintainability a fair bit? Would also lead `otel_span` to be something like `{ .span_id, .trace_id, .span }` which also feels like a simpler API?

Reply:
Yup, it makes for a nice simplification in 0c124ec496fc15c77ce85d44b18d435fcfc827e0 (PR #3987)